### PR TITLE
Lever hiring oauth backend

### LINKED ISSUE #7110: oAuth backend for lever-hiring source
Part of #7000 story

Implement lever-hiring  oauth backend and do e2e test.

### PR DESCRIPTION
**Pull request by eliziario**
## What
implements https://github.com/airbytehq/airbyte/issues/7110 

## How
Implement the AuthFlow class in the backend and update the spec with the correct fields for oauth authentication

## Recommended reading order
1. spec.json
2. source.py
3.LeverOAuthFlow.java

## Pre-merge Checklist

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [X] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

### CODE REVIEW

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            .put("redirect_uri", redirectUrl)
+            .put("client_secret", clientSecret)
+            .put("grant_type", "authorization_code")
+            .put("code", authCode)
+            .build();
+  }
+
+  @Override
+  protected Map<String, Object> extractRefreshToken(JsonNode data, String accessTokenUrl) throws IOException {
+    System.out.println(Jsons.serialize(data));
```
Comment (ChristopheDuong) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java:54]: Can we avoid having System.out code in PR please?

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+      return Map.of("refresh_token", refreshToken);
+    } else {
+      throw new IOException(String.format("Missing 'refresh_token' in query params from %s", accessTokenUrl));
+    }
+  }
+
+
+  @Override
+  protected String formatConsentUrl(UUID definitionId, String clientId, String redirectUrl) throws IOException {
+    return String.format("%s?client_id=%s&redirect_uri=%s&response_type=code&state=%s&scope=%s&prompt=consent&audience=%s",
```
Comment (ChristopheDuong) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java:66]: why do you not use the `URIBuilder` like the other implementations?

**Review by ChristopheDuong — CHANGES_REQUESTED** [review]

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import java.io.IOException;
+import java.net.URISyntaxException;
+import java.net.http.HttpClient;
+import java.util.Map;
+import java.util.UUID;
+import java.util.function.Supplier;
+
+public class LeverOAuthFlow extends BaseOAuthFlow {
+
+  private static final String AUTHORIZE_URL = "https://sandbox-lever.auth0.com/authorize";
```
Comment (sherifnada) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java:24]: This URL is only for sandbox no? It wouldn't work for a customers actual account

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      "postings:read:admin",
+      "referrals:read:admin",
+      "requisitions:read:admin",
+      "resumes:read:admin",
+      "sources:read:admin",
+      "stages:read:admin",
+      "offline_access");
+
+  private String getAudience() {
+    return "https://api.sandbox.lever.co/v1/";
```
Comment (sherifnada) [inline airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java:41]: Sandbox

**Review by sherifnada — CHANGES_REQUESTED** [review]
It seems we are pointing to sandbox URLs. Also Shouldn't it also be added to the flow factory?

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java (hunk 3 above)
Comment (eliziario) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java:24]: We need to schedule a meeting with them so they do QA in our integration, where we would demo it live against the sandbox environment. Will discuss this with Karina.

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java (hunk 4 above)
Comment (eliziario) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java:41]: See other comment. Lever's application approval is needed.

File: airbyte-oauth/src/test/java/io/airbyte/oauth/flows/LeverOAuthFlowTest.java [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+import java.io.IOException;
+import java.net.http.HttpClient;
+import java.net.http.HttpResponse;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class LeverOAuthFlowTest {
```
Comment (ChristopheDuong) [inline airbyte-oauth/src/test/java/io/airbyte/oauth/flows/LeverOAuthFlowTest.java:27]: please extends `BaseOAuthFlowTest` instead

**Review by ChristopheDuong — CHANGES_REQUESTED** [review]

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java (hunk 3 above)
Comment (eliziario) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java]: fixed

File: airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java (hunk 4 above)
Comment (eliziario) [reply airbyte-oauth/src/main/java/io/airbyte/oauth/flows/LeverOAuthFlow.java]: fixed

**Review by sherifnada — APPROVED** [review]

**Comment by eliziario** [comment]
test connector=connectors/source-lever-hiring

**Pull request merged**